### PR TITLE
[webview_flutter] Add a backgroundColor option to the iOS webview

### PR DESCRIPTION
**Pull request by e-adrien**
This PR add an option to set the background color of the iOS webview.

Original PR :

https://github.com/flutter/plugins/pull/3431

It fixes this issue :

https://github.com/flutter/flutter/issues/29300

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m [hunk 1]
```diff
@@ -96,6 +96,20 @@ - (instancetype)initWithFrame:(CGRect)frame
...
     _webView = [[FLTWKWebView alloc] initWithFrame:frame configuration:configuration];
+
+    // Background color
+    NSNumber* backgroundColorNSNumber = args[@"backgroundColor"];
+    if ([backgroundColorNSNumber isKindOfClass:[NSNumber class]]) {
+      int backgroundColorInt = [backgroundColorNSNumber intValue];
+      UIColor* backgroundColor = [UIColor colorWithRed:(backgroundColorInt >> 16 & 0xff) / 255.f
+                                                 green:(backgroundColorInt >> 8 & 0xff) / 255.f
+                                                  blue:(backgroundColorInt & 0xff) / 255.f
+                                                 alpha:(backgroundColorInt >> 24 & 0xff) / 255.f];
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m:107]: Nit: CGFloat is essentially always 64-bit these days, so these don't need to be floats: `/ 255.0`

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m [hunk 2]
```diff
@@ -96,6 +96,20 @@ - (instancetype)initWithFrame:(CGRect)frame
...
+      _webView.opaque = NO;
+      _webView.backgroundColor = UIColor.clearColor;
+      _webView.scrollView.backgroundColor = backgroundColor;
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m:110]: Non-opaque views have worse performance, so it's almost certainly worth special-casing this:
```
int alpha = backgroundColorInt >> 24 & 0xff;
UIColor* backgroundColor = [... alpha:(alpha / 255.0)];
if (alpha == 255) {
  _webView.backgroundColor = backgroundColor;
} else {
  [these three lines]
}
```

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looks great other than minor nits.

/cc @jmagman or @mvanbeusekom for secondary review.

File: packages/webview_flutter/webview_flutter_wkwebview/example/lib/web_view.dart [hunk 3]
```diff
@@ -227,6 +228,13 @@ class WebView extends StatefulWidget {
   /// The default policy is [AutoMediaPlaybackPolicy.require_user_action_for_all_media_types].
   final AutoMediaPlaybackPolicy initialMediaPlaybackPolicy;
 
+  /// The background color of the webview.
```
Comment (mvanbeusekom) [inline packages/webview_flutter/webview_flutter_wkwebview/example/lib/web_view.dart:231]: nit: convert "webview" to breadcrum:
```suggestion
  /// The background color of the [WebView].
```

File: packages/webview_flutter/webview_flutter_wkwebview/example/lib/web_view.dart [hunk 4]
```diff
@@ -227,6 +228,13 @@ class WebView extends StatefulWidget {
...
+  ///
+  /// When null the platform's webview default background color is used.
+  ///
+  /// By default `backgroundColor` is null.
```
Comment (mvanbeusekom) [inline packages/webview_flutter/webview_flutter_wkwebview/example/lib/web_view.dart:235]: nit: Could be combined into one paragraph:
```suggestion
  /// When `null` the platform's webview default background color is used. By 
  /// default [backgroundColor] is `null`.
```

**Review by mvanbeusekom — APPROVED** [review]
LGTM apart from comments Stuart already made and two small additional nits.

File: packages/webview_flutter/webview_flutter_wkwebview/example/lib/main.dart [hunk 5]
```diff
@@ -70,6 +91,7 @@ class _WebViewExampleState extends State<_WebViewExample> {
   @override
   Widget build(BuildContext context) {
     return Scaffold(
+      backgroundColor: Colors.green,
```
Comment (jmagman) [inline packages/webview_flutter/webview_flutter_wkwebview/example/lib/main.dart:94]: What about hard coding the color so if it changes in the framework it doesn't break this test in the plugin repo only on the master tests?
```suggestion
      backgroundColor: Color(0xFF4CAF50),
```

**Review by jmagman — APPROVED** [review]
Looks good once nits and opaque color assignment special casing added.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m (hunk 2 above)
Comment (e-adrien) [reply packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m:110]: I tried to make the change, but if I don't set `_webView.opaque = NO;` for an opaque color too the WebView's background is white on my emulator.

Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m:110]: That's unfortunate; hopefully that means the drawing for WKWebView is doing something non-obvious then which will avoid the performance penalty.

**Review by stuartmorgan — APPROVED** [review]
LGTM!

**Pull request merged**